File: src/db/task_store.go
// src/db/task_store.go
package db

import (
	"database/sql"
	"fmt"
	"strings"
	"time"

	"github.com/lucas/oraculo/src/domain"
)

// TaskStore provides CRUD operations and lifecycle transitions for tasks backed by SQLite.
// It includes DAG validation to prevent cyclic dependencies.
type TaskStore struct {
	db *DB
}

// NewTaskStore returns a TaskStore that uses the given database connection.
func NewTaskStore(db *DB) *TaskStore {
	return &TaskStore{db: db}
}

// scanTask reads a Task from a row scanner, parsing SQLite TEXT timestamps.
// started_at and completed_at may be NULL.
func scanTask(row interface{ Scan(...any) error }) (*domain.Task, error) {
	var (
		t                    domain.Task
		createdAt, updatedAt string
		startedAt            *string
		completedAt          *string
	)
	if err := row.Scan(
		&t.ID, &t.StoryID, &t.Name, &t.Description, &t.Status,
		&t.FailureReason, &createdAt, &updatedAt, &startedAt, &completedAt,
	); err != nil {
		return nil, err
	}
	var err error
	if t.CreatedAt, err = time.Parse(sqliteTimeFmt, createdAt); err != nil {
		return nil, fmt.Errorf("parse created_at: %w", err)
	}
	if t.UpdatedAt, err = time.Parse(sqliteTimeFmt, updatedAt); err != nil {
		return nil, fmt.Errorf("parse updated_at: %w", err)
	}
	if startedAt != nil {
		parsed, err := time.Parse(sqliteTimeFmt, *startedAt)
		if err != nil {
			return nil, fmt.Errorf("parse started_at: %w", err)
		}
		t.StartedAt = &parsed
	}
	if completedAt != nil {
		parsed, err := time.Parse(sqliteTimeFmt, *completedAt)
		if err != nil {
			return nil, fmt.Errorf("parse completed_at: %w", err)
		}
		t.CompletedAt = &parsed
	}
	return &t, nil
}

const taskColumns = "id, story_id, name, description, status, failure_reason, created_at, updated_at, started_at, completed_at"

// Create inserts a new task or returns the existing one if (storyID, name) already exists.
// The created flag is true when a new row was inserted.
// If dependsOn is non-empty, dependency edges are created after resolving names to task IDs
// within the same story. Cycle detection is performed before inserting edges.
func (s *TaskStore) Create(storyID int, name, description string, dependsOn []string) (*domain.Task, bool, error) {
	// Use a transaction so that the insert + dependency edges are atomic.
	tx, err := s.db.conn.Begin()
	if err != nil {
		return nil, false, fmt.Errorf("begin tx: %w", err)
	}
	defer tx.Rollback()

	res, err := tx.Exec(
		"INSERT OR IGNORE INTO tasks (story_id, name, description) VALUES (?, ?, ?)",
		storyID, name, description,
	)
	if err != nil {
		return nil, false, fmt.Errorf("create task: %w", err)
	}
	affected, err := res.RowsAffected()
	if err != nil {
		return nil, false, fmt.Errorf("rows affected: %w", err)
	}

	// Fetch the task (whether newly created or existing).
	row := tx.QueryRow(
		"SELECT "+taskColumns+" FROM tasks WHERE story_id = ? AND name = ?",
		storyID, name,
	)
	task, err := scanTask(row)
	if err != nil {
		return nil, false, fmt.Errorf("fetch after create: %w", err)
	}

	// Process dependencies if any.
	if len(dependsOn) > 0 {
		if err := s.addDependencies(tx, storyID, task.ID, name, dependsOn); err != nil {
			return nil, false, err
		}
	}

	if err := tx.Commit(); err != nil {
		return nil, false, fmt.Errorf("commit: %w", err)
	}

	return task, affected > 0, nil
}

// addDependencies resolves dependency names to task IDs and inserts edges,
// performing cycle detection before each insertion.
func (s *TaskStore) addDependencies(tx *sql.Tx, storyID, taskID int, taskName string, dependsOn []string) error {
	for _, depName := range dependsOn {
		// Self-dependency check (also enforced by CHECK constraint).
		if depName == taskName {
			return fmt.Errorf("task %q cannot depend on itself: %w", taskName, domain.ErrCyclicDependency)
		}

		// Resolve dependency name to ID within the same story.
		var depID int
		err := tx.QueryRow(
			"SELECT id FROM tasks WHERE story_id = ? AND name = ?",
			storyID, depName,
		).Scan(&depID)
		if err != nil {
			if err == sql.ErrNoRows {
				return fmt.Errorf("dependency %q: %w", depName, domain.ErrNotFound)
			}
			return fmt.Errorf("resolve dependency %q: %w", depName, err)
		}

		// Cycle detection: check if adding (taskID depends on depID) would create a cycle.
		// A cycle exists if taskID is reachable from depID by following "depends on" edges.
		if err := s.detectCycle(tx, taskID, depID); err != nil {
			return err
		}

		// Insert the dependency edge. Use INSERT OR IGNORE to be idempotent.
		_, err = tx.Exec(
			"INSERT OR IGNORE INTO task_dependencies (task_id, depends_on) VALUES (?, ?)",
			taskID, depID,
		)
		if err != nil {
			return fmt.Errorf("insert dependency %q -> %q: %w", taskName, depName, err)
		}
	}
	return nil
}

// detectCycle checks whether adding an edge (taskID depends on depID) would create a cycle.
// It does BFS from depID following "depends on" edges: for each node X, find all nodes that
// X depends on (SELECT depends_on FROM task_dependencies WHERE task_id = X).
// If we reach taskID, there is a cycle.
func (s *TaskStore) detectCycle(tx *sql.Tx, taskID, depID int) error {
	visited := map[int]bool{}
	queue := []int{depID}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if current == taskID {
			return fmt.Errorf("adding dependency would create cycle: %w", domain.ErrCyclicDependency)
		}

		if visited[current] {
			continue
		}
		visited[current] = true

		rows, err := tx.Query(
			"SELECT depends_on FROM task_dependencies WHERE task_id = ?", current,
		)
		if err != nil {
			return fmt.Errorf("query dependencies: %w", err)
		}

		for rows.Next() {
			var next int
			if err := rows.Scan(&next); err != nil {
				rows.Close()
				return fmt.Errorf("scan dependency: %w", err)
			}
			if !visited[next] {
				queue = append(queue, next)
			}
		}
		rows.Close()
		if err := rows.Err(); err != nil {
			return fmt.Errorf("iterate dependencies: %w", err)
		}
	}
	return nil
}

// Start transitions a task from pending to in_progress, setting started_at.
// It returns domain.ErrInvalidTransition if the task is not in pending status.
func (s *TaskStore) Start(storyID int, name string) (*domain.Task, error) {
	task, err := s.GetByName(storyID, name)
	if err != nil {
		return nil, err
	}

	if !task.Status.CanTransitionTo(domain.TaskInProgress) {
		return nil, fmt.Errorf("task %q: cannot transition from %s to in_progress: %w",
			name, task.Status, domain.ErrInvalidTransition)
	}

	_, err = s.db.conn.Exec(
		"UPDATE tasks SET status = ?, started_at = datetime('now'), updated_at = datetime('now') WHERE story_id = ? AND name = ?",
		domain.TaskInProgress, storyID, name,
	)
	if err != nil {
		return nil, fmt.Errorf("start task: %w", err)
	}

	return s.GetByName(storyID, name)
}

// Complete transitions a task from in_progress to completed, setting completed_at
// and inserting a TaskResult row.
// It returns domain.ErrInvalidTransition if the task is not in in_progress status.
func (s *TaskStore) Complete(storyID int, name string, result domain.TaskResult) (*domain.Task, error) {
	task, err := s.GetByName(storyID, name)
	if err != nil {
		return nil, err
	}

	if !task.Status.CanTransitionTo(domain.TaskCompleted) {
		return nil, fmt.Errorf("task %q: cannot transition from %s to completed: %w",
			name, task.Status, domain.ErrInvalidTransition)
	}

	tx, err := s.db.conn.Begin()
	if err != nil {
		return nil, fmt.Errorf("begin tx: %w", err)
	}
	defer tx.Rollback()

	_, err = tx.Exec(
		"UPDATE tasks SET status = ?, completed_at = datetime('now'), updated_at = datetime('now') WHERE story_id = ? AND name = ?",
		domain.TaskCompleted, storyID, name,
	)
	if err != nil {
		return nil, fmt.Errorf("complete task: %w", err)
	}

	skillsUsed := strings.Join(result.SkillsUsed, ",")
	filesModified := strings.Join(result.FilesModified, ",")

	_, err = tx.Exec(
		"INSERT INTO task_results (task_id, summary, logs, skills_used, files_modified) VALUES (?, ?, ?, ?, ?)",
		task.ID, result.Summary, result.Logs, skillsUsed, filesModified,
	)
	if err != nil {
		return nil, fmt.Errorf("insert task result: %w", err)
	}

	if err := tx.Commit(); err != nil {
		return nil, fmt.Errorf("commit: %w", err)
	}

	return s.GetByName(storyID, name)
}

// Fail transitions a task from in_progress to failed, setting failure_reason.
// It returns domain.ErrInvalidTransition if the task is not in in_progress status.
func (s *TaskStore) Fail(storyID int, name, reason string) (*domain.Task, error) {
	task, err := s.GetByName(storyID, name)
	if err != nil {
		return nil, err
	}

	if !task.Status.CanTransitionTo(domain.TaskFailed) {
		return nil, fmt.Errorf("task %q: cannot transition from %s to failed: %w",
			name, task.Status, domain.ErrInvalidTransition)
	}

	_, err = s.db.conn.Exec(
		"UPDATE tasks SET status = ?, failure_reason = ?, updated_at = datetime('now') WHERE story_id = ? AND name = ?",
		domain.TaskFailed, reason, storyID, name,
	)
	if err != nil {
		return nil, fmt.Errorf("fail task: %w", err)
	}

	return s.GetByName(storyID, name)
}

// GetByName retrieves a task by its unique (storyID, name) pair.
// It returns domain.ErrNotFound when no matching task exists.
func (s *TaskStore) GetByName(storyID int, name string) (*domain.Task, error) {
	row := s.db.conn.QueryRow(
		"SELECT "+taskColumns+" FROM tasks WHERE story_id = ? AND name = ?",
		storyID, name,
	)
	task, err := scanTask(row)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, fmt.Errorf("task %q (story %d): %w", name, storyID, domain.ErrNotFound)
		}
		return nil, fmt.Errorf("get task: %w", err)
	}
	return task, nil
}

// List returns all tasks for the given storyID ordered by creation time.
func (s *TaskStore) List(storyID int) ([]domain.Task, error) {
	rows, err := s.db.conn.Query(
		"SELECT "+taskColumns+" FROM tasks WHERE story_id = ? ORDER BY created_at",
		storyID,
	)
	if err != nil {
		return nil, fmt.Errorf("list tasks: %w", err)
	}
	defer rows.Close()

	var tasks []domain.Task
	for rows.Next() {
		task, err := scanTask(rows)
		if err != nil {
			return nil, fmt.Errorf("scan task: %w", err)
		}
		tasks = append(tasks, *task)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate tasks: %w", err)
	}
	return tasks, nil
}

// Delete removes a task by (storyID, name).
// It returns domain.ErrNotFound when no matching task exists.
func (s *TaskStore) Delete(storyID int, name string) error {
	res, err := s.db.conn.Exec(
		"DELETE FROM tasks WHERE story_id = ? AND name = ?",
		storyID, name,
	)
	if err != nil {
		return fmt.Errorf("delete task: %w", err)
	}
	affected, err := res.RowsAffected()
	if err != nil {
		return fmt.Errorf("rows affected: %w", err)
	}
	if affected == 0 {
		return fmt.Errorf("task %q (story %d): %w", name, storyID, domain.ErrNotFound)
	}
	return nil
}

File: apps/orchestrator/src/db/stores/task-store.ts
import type { Database } from "bun:sqlite";
import type { Task, TaskEnriched, TaskResult } from "../../domain/entities.ts";
import type { TaskStatus } from "../../domain/enums.ts";
import {
  ErrNotFound,
  ErrCyclicDependency,
  ErrInvalidTransition,
} from "../../domain/errors.ts";
import { canTransition } from "../../domain/state.ts";

const TASK_COLUMNS =
  "id, story_id, name, description, status, failure_reason, created_at, updated_at, started_at, completed_at";

interface TaskRow {
  id: number;
  story_id: number;
  name: string;
  description: string;
  status: TaskStatus;
  failure_reason: string;
  created_at: string;
  updated_at: string;
  started_at: string | null;
  completed_at: string | null;
}

function rowToTask(row: TaskRow): Task {
  const task: Task = {
    id: row.id,
    story_id: row.story_id,
    name: row.name,
    description: row.description,
    status: row.status,
    failure_reason: row.failure_reason,
    created_at: row.created_at,
    updated_at: row.updated_at,
  };
  if (row.started_at != null) task.started_at = row.started_at;
  if (row.completed_at != null) task.completed_at = row.completed_at;
  return task;
}

export interface CompleteInput {
  summary: string;
  logs: string;
  skills_used: string[];
  files_modified: string[];
}

export class TaskStore {
  constructor(private db: Database) {}

  /**
   * Inserts a new task or returns the existing one if (storyId, name) already exists.
   * Returns [task, created] where created is true when a new row was inserted.
   * If dependsOn is non-empty, dependency edges are created atomically.
   */
  create(
    storyId: number,
    name: string,
    description: string,
    dependsOn: string[],
  ): [Task, boolean] {
    // Use a transaction so insert + dependency edges are atomic
    const tx = this.db.transaction(() => {
      const res = this.db.run(
        "INSERT OR IGNORE INTO tasks (story_id, name, description) VALUES (?, ?, ?)",
        [storyId, name, description],
      );
      const affected = res.changes;

      const row = this.db
        .query(`SELECT ${TASK_COLUMNS} FROM tasks WHERE story_id = ? AND name = ?`)
        .get(storyId, name) as TaskRow;

      if (dependsOn.length > 0) {
        this.addDependencies(storyId, row.id, name, dependsOn);
      }

      return [rowToTask(row), affected > 0] as [Task, boolean];
    });

    return tx();
  }

  /**
   * Retrieves a task by its unique (storyId, name) pair.
   * Throws ErrNotFound when no matching task exists.
   */
  getByName(storyId: number, name: string): Task {
    const row = this.db
      .query(`SELECT ${TASK_COLUMNS} FROM tasks WHERE story_id = ? AND name = ?`)
      .get(storyId, name) as TaskRow | null;

    if (!row) {
      throw ErrNotFound("task", name);
    }
    return rowToTask(row);
  }

  /**
   * Returns all tasks for the given storyId ordered by creation time.
   */
  list(storyId: number): Task[] {
    const rows = this.db
      .query(
        `SELECT ${TASK_COLUMNS} FROM tasks WHERE story_id = ? ORDER BY created_at`,
      )
      .all(storyId) as TaskRow[];

    return rows.map(rowToTask);
  }

  /**
   * Returns all tasks for the given storyId with dependencies and results hydrated.
   */
  listEnriched(storyId: number): TaskEnriched[] {
    const tasks = this.list(storyId);
    if (tasks.length === 0) return [];

    const ids = tasks.map((t) => t.id);
    const idxById = new Map<number, number>();
    tasks.forEach((t, i) => idxById.set(t.id, i));

    const enriched: TaskEnriched[] = tasks.map((t) => ({
      ...t,
      depends_on: [],
    }));

    // Batch fetch dependencies
    const placeholders = ids.map(() => "?").join(",");
    const depRows = this.db
      .query(
        `SELECT task_id, depends_on FROM task_dependencies WHERE task_id IN (${placeholders})`,
      )
      .all(...ids) as { task_id: number; depends_on: number }[];

    for (const row of depRows) {
      const idx = idxById.get(row.task_id);
      if (idx !== undefined) {
        enriched[idx].depends_on.push(row.depends_on);
      }
    }

    // Batch fetch results
    const resRows = this.db
      .query(
        `SELECT id, task_id, summary, logs, skills_used, files_modified, created_at FROM task_results WHERE task_id IN (${placeholders})`,
      )
      .all(...ids) as {
      id: number;
      task_id: number;
      summary: string;
      logs: string;
      skills_used: string;
      files_modified: string;
      created_at: string;
    }[];

    for (const row of resRows) {
      const idx = idxById.get(row.task_id);
      if (idx !== undefined) {
        enriched[idx].result = {
          id: row.id,
          task_id: row.task_id,
          summary: row.summary,
          logs: row.logs,
          skills_used: row.skills_used ? row.skills_used.split(",") : [],
          files_modified: row.files_modified
            ? row.files_modified.split(",")
            : [],
          created_at: row.created_at,
        };
      }
    }

    return enriched;
  }

  /**
   * Transitions a task from pending to in_progress, setting started_at.
   */
  start(storyId: number, name: string): Task {
    const task = this.getByName(storyId, name);

    if (!canTransition(task.status, "in_progress")) {
      throw ErrInvalidTransition(task.status, "in_progress");
    }

    this.db.run(
      "UPDATE tasks SET status = ?, started_at = datetime('now'), updated_at = datetime('now') WHERE story_id = ? AND name = ?",
      ["in_progress", storyId, name],
    );

    return this.getByName(storyId, name);
  }

  /**
   * Transitions a task from in_progress to completed and inserts a TaskResult.
   */
  complete(storyId: number, name: string, input: CompleteInput): Task {
    const task = this.getByName(storyId, name);

    if (!canTransition(task.status, "completed")) {
      throw ErrInvalidTransition(task.status, "completed");
    }

    const tx = this.db.transaction(() => {
      this.db.run(
        "UPDATE tasks SET status = ?, completed_at = datetime('now'), updated_at = datetime('now') WHERE story_id = ? AND name = ?",
        ["completed", storyId, name],
      );

      const skillsUsed = input.skills_used.join(",");
      const filesModified = input.files_modified.join(",");

      this.db.run(
        "INSERT INTO task_results (task_id, summary, logs, skills_used, files_modified) VALUES (?, ?, ?, ?, ?)",
        [task.id, input.summary, input.logs, skillsUsed, filesModified],
      );
    });

    tx();
    return this.getByName(storyId, name);
  }

  /**
   * Transitions a task from in_progress to failed, setting failure_reason.
   */
  fail(storyId: number, name: string, reason: string): Task {
    const task = this.getByName(storyId, name);

    if (!canTransition(task.status, "failed")) {
      throw ErrInvalidTransition(task.status, "failed");
    }

    this.db.run(
      "UPDATE tasks SET status = ?, failure_reason = ?, updated_at = datetime('now') WHERE story_id = ? AND name = ?",
      ["failed", reason, storyId, name],
    );

    return this.getByName(storyId, name);
  }

  /**
   * Deletes a task by (storyId, name).
   * Throws ErrNotFound when no matching task exists.
   */
  delete(storyId: number, name: string): void {
    const res = this.db.run(
      "DELETE FROM tasks WHERE story_id = ? AND name = ?",
      [storyId, name],
    );
    if (res.changes === 0) {
      throw ErrNotFound("task", name);
    }
  }

  // --- Private helpers ---

  /**
   * Resolves dependency names to task IDs and inserts edges,
   * performing cycle detection before each insertion.
   */
  private addDependencies(
    storyId: number,
    taskId: number,
    taskName: string,
    dependsOn: string[],
  ): void {
    for (const depName of dependsOn) {
      // Self-dependency check
      if (depName === taskName) {
        throw ErrCyclicDependency(taskName, depName);
      }

      // Resolve dependency name to ID within the same story
      const depRow = this.db
        .query("SELECT id FROM tasks WHERE story_id = ? AND name = ?")
        .get(storyId, depName) as { id: number } | null;

      if (!depRow) {
        throw ErrNotFound("task dependency", depName);
      }

      // Cycle detection
      this.detectCycle(taskId, depRow.id);

      // Insert the dependency edge (idempotent)
      this.db.run(
        "INSERT OR IGNORE INTO task_dependencies (task_id, depends_on) VALUES (?, ?)",
        [taskId, depRow.id],
      );
    }
  }

  /**
   * BFS cycle detection: checks whether adding an edge (taskId depends on depId)
   * would create a cycle. A cycle exists if taskId is reachable from depId
   * by following "depends on" edges.
   */
  private detectCycle(taskId: number, depId: number): void {
    const visited = new Set<number>();
    const queue: number[] = [depId];

    while (queue.length > 0) {
      const current = queue.shift()!;

      if (current === taskId) {
        throw ErrCyclicDependency(taskId, depId);
      }

      if (visited.has(current)) continue;
      visited.add(current);

      const rows = this.db
        .query("SELECT depends_on FROM task_dependencies WHERE task_id = ?")
        .all(current) as { depends_on: number }[];

      for (const row of rows) {
        if (!visited.has(row.depends_on)) {
          queue.push(row.depends_on);
        }
      }
    }
  }
}
